src/common/query/src/logical_plan.rs

```diff
@@ -68,6 +71,21 @@ pub fn create_aggregate_function(
     )
 }
 
+/// The datafusion `[LogicalPlan]` decoder.
+#[async_trait::async_trait]
+pub trait SubstraitPlanDecoder {
+    /// Decode the `[LogicalPlan]` from bytes with the `[CatalogProviderList]`.
+    /// When `optimize` is true, it will do the optimization for decoded plan.
+    async fn decode(
+        &self,
+        message: bytes::Bytes,
+        catalog_list: Arc<dyn CatalogProviderList>,
+        optimize: bool,
```

Comment:
Why a decoder needs to optimize the plan? We prefer to remove this param and let others call optimizer by themselves

Reply:
Agree, let me fix it.

Reply:
I found the reason why I did it this way. The reason is that we need to decode the plan in `DfTableSourceProvider`, which is in the `catalog` crate. 

However, we can't call the optimizer in the `catalog` crate because the `SessionState` is only in the `query` crate. If we let the `query` catalog depend on the `catalog`, there will be a cyclic dependency issue.

Reply:
I leave a `TODO` here. If you have a great suggestion, I am open to refactoring it.